gl.cpp

```diff
@@ -354,29 +370,36 @@ GLFix nglZBufferAt(const unsigned int x, const unsigned int y)
     return z_buffer[x + y * SCREEN_WIDTH];
 }
 
+constexpr GLFix ABS(GLFix a) {
+    return a >= GLFix(0) ? a : -a;
+}
+
 //Doesn't interpolate colors even if enabled
 void nglDrawLine3D(const VERTEX *v1, const VERTEX *v2)
 {    
-    //TODO: Z-Clipping
+    // Z-Clipping!
+    if(v1->z < near_plane || v2->z < near_plane)
+        return;
 
     VERTEX v1_p = *v1, v2_p = *v2;
     nglPerspective(&v1_p);
     nglPerspective(&v2_p);
 
     const GLFix diff_x = v2_p.x - v1_p.x;
-    const GLFix dy = (v2_p.y - v1_p.y) / diff_x;
+    const GLFix diff_y = v2_p.y - v1_p.y;
 
     const COLOR c = v1_p.c;
 
     //Height > width? -> Interpolate X
-    if(dy > GLFix(1) || dy < GLFix(-1))
+    //if(dy > GLFix(1) || dy < GLFix(-1))
+
+    // if check passes diff_y should always be non zero
+    if (ABS(diff_x) < ABS(diff_y))
     {
         if(v2_p.y < v1_p.y)
             std::swap(v1_p, v2_p);
 
-        const GLFix diff_y = v2_p.y - v1_p.y;
-
-        const GLFix dx = (v2_p.x - v1_p.x) / diff_y;
+        const GLFix dx = diff_x / diff_y;
```

Comment:
I don't think it's correct to use those values after the `std::swap` above. The swap is there to guarantee that `v1_p` has a lower Y than `v2_p`, so you would have to change the sign of all the `diff_` values as well. Otherwise drawn lines face the wrong direction.

It's easier to just always compute the diffs after the swap though - subtraction is cheap.

Reply:
Using `diff_x / diff_y` is fine here. In the case where the points have to be swapped, `diff_y` is negative and the division returns the correct result. It's important to not mix diffs before and after the swap though, which is the case in line 403 for instance.